init/caddy.service

```diff
@@ -20,6 +20,8 @@ After=network.target network-online.target
 Requires=network-online.target
 
 [Service]
+Type=notify
+NotifyAccess=exec
```

Comment:
Is it necessary to set `NotifyAccess` to `exec`?  When `Type` is `notify`, `NotifyAccess` defaults to `main`, meaning service updates are only accepted from the main process.  Caddy just has the one process, correct?  Using `exec` is only needed if you have a separate control process.

- https://www.freedesktop.org/software/systemd/man/systemd.service.html#Type=
- https://www.freedesktop.org/software/systemd/man/systemd.service.html#NotifyAccess=

Reply:
As far as I understand this, `NotifyAccess=exec` is required for `caddy reload`. This is triggered via an external command.

Reply:
Why?  sd_notify is for signaling that the process has finished starting up.  `caddy reload` doesn't restart the process.

Reply:
We're also notifying of reloading @carlwgeorge https://github.com/caddyserver/caddy/pull/3963/files#diff-46e2b06a46e2ad70a768686e9d95500a1f5c10b76538d3150e2e9947855694c4R64 (but don't take that as a disagreement, you know more about this than I do)

Reply:
Ah, I thought sd_notify was exclusive to startup, I didn't realize it was also used for other events.  Thanks for the additional context.

Whether or not we need `NotifyAccess=exec` depends on which process is sending the signal to the notify socket.  Which of these is accurate?

* `caddy reload` process -> `caddy run` process -> notify socket

* `caddy reload` process -> `caddy run` process
  `caddy reload` process -> notify socket

Reply:
`caddy reload` sends an API call to the actively running Caddy instance (usually over TCP, but could be over unix socket, doesn't matter) and then running Caddy instance will take that given config and attempt a reload. So it is the `caddy run` process that is sending the notify. So `main` sounds correct to me, I think.

Reply:
Really?

https://github.com/rumpelsepp/caddy/blob/6a018a33314386c2658c4810a0c738d3e231ebad/cmd/commandfuncs.go#L298

The mentioned API request is issued in `cmdReload()` [here](https://github.com/rumpelsepp/caddy/blob/6a018a33314386c2658c4810a0c738d3e231ebad/cmd/commandfuncs.go#L330). Thus I assume that the `caddy reload` process issues the notification.

Reply:
Ah you're right, misread the code.

Reply:
That's why the setting is here. systemd does some magic with unit scope, cgroups, other metadata from the service file… Hence, systemd knows that the **service** did a reload. AFAIU it doesn'matter which process notifies systemd. The `NotifyAccess` setting is about permissions.

Reply:
The only difference is if the `caddy run` process sends the signal, we don't need to set the `NotifyAccess` directive at all.  That would be my preference to keep the unit file as small as possible, but I don't know how hard it would be to change the code match that behavior.  As the code is now, these directives are correct.

Reply:
I'm unclear on what the value of notifying of reloads even is, cause Caddy remains functional during reloads, it does them gracefully by running the old and new configs at the same time for a short period of time before turning off the old config. Is it just so you see the status as "Reloading" momentarily? Seems to be of little value to me but :man_shrugging: 

As-is, systemd would not be alerted of config changes made directly via the API via curl or via running `caddy reload --config /etc/caddy/Caddyfile` as a regular user instead of via `systemctl reload caddy`. Does that matter?

Reply:
> Is it just so you see the status as "Reloading" momentarily?

Exactly.  From the [docs](https://www.freedesktop.org/software/systemd/man/sd_notify.html#RELOADING=1): "This is useful to allow the service manager to track the service's internal state, and present it to the user."

> Does that matter?

Yes, to the extent it matters to notify for reloads at all.  As you said it would be a very brief window where the user would see that status reflected.

Reply:
And a long window might indicate a problem (or misconfiguration).

Reply:
So what's the plan with this?

Reply:
I think we can merge this, Francis and I think it seems harmless enough. If there's no objections, let's do it in the next day or so.

Reply:
Heads up @rumpelseep, I submitted https://github.com/caddyserver/caddy/pull/4060 to address the concerns I had.  If that is merged then we won't need `NotifyAccess=exec` in the unit file.